src/ray/gcs/gcs_server/gcs_table_storage.h

```diff
@@ -0,0 +1,382 @@
+
+  Status Get(const Key &key, const OptionalItemCallback<Data> &callback);
+
+  Status GetAll(const SegmentedCallback<std::pair<Key, Data>> &callback);
+
+  Status Delete(const Key &key, const StatusCallback &callback);
+
+  std::string table_name_;
+  std::shared_ptr<StoreClient<Key, Data, JobID>> store_client_;
+};
+
+/// \class GcsTableWithIndex
+///
+/// GcsTableWithIndex supports putting, getting and deleting of GCS table data.
+/// This class is not meant to be used directly. All gcs table classes with job id should
+/// derive from this class and override the table_name_ member with a unique value
+/// for that table.
+template <typename Key, typename Data>
+class GcsTableWithIndex : public GcsTable<Key, Data> {
+ public:
+  explicit GcsTableWithIndex(std::shared_ptr<StoreClient<Key, Data, JobID>> store_client)
+      : GcsTable<Key, Data>(store_client) {}
+
+  Status PutWithJobId(const JobID &job_id, const Key &key, const Data &value,
+                      const StatusCallback &callback);
+
+  Status GetByJobId(const JobID &job_id,
+                    const SegmentedCallback<std::pair<Key, Data>> &callback);
+
+  Status DeleteByJobId(const JobID &job_id, const StatusCallback &callback);
+};
+
+class GcsJobTable : public GcsTable<JobID, JobTableData> {
+ public:
+  explicit GcsJobTable(
+      std::shared_ptr<StoreClient<JobID, JobTableData, JobID>> store_client)
+      : GcsTable(std::move(store_client)) {
+    table_name_ = TablePrefix_Name(TablePrefix::JOB);
+  }
+};
+
+class GcsActorTable : public GcsTableWithIndex<ActorID, ActorTableData> {
+ public:
+  explicit GcsActorTable(
+      std::shared_ptr<StoreClient<ActorID, ActorTableData, JobID>> store_client)
+      : GcsTableWithIndex(std::move(store_client)) {
+    table_name_ = TablePrefix_Name(TablePrefix::ACTOR);
+  }
+};
+
+class GcsActorCheckpointTable
+    : public GcsTableWithIndex<ActorCheckpointID, ActorCheckpointData> {
+ public:
+  explicit GcsActorCheckpointTable(
+      std::shared_ptr<StoreClient<ActorCheckpointID, ActorCheckpointData, JobID>>
+          store_client)
+      : GcsTableWithIndex(std::move(store_client)) {
+    table_name_ = TablePrefix_Name(TablePrefix::ACTOR_CHECKPOINT);
+  }
+};
+
+class GcsActorCheckpointIdTable
+    : public GcsTableWithIndex<ActorID, ActorCheckpointIdData> {
+ public:
+  explicit GcsActorCheckpointIdTable(
+      std::shared_ptr<StoreClient<ActorID, ActorCheckpointIdData, JobID>> store_client)
+      : GcsTableWithIndex(std::move(store_client)) {
+    table_name_ = TablePrefix_Name(TablePrefix::ACTOR_CHECKPOINT_ID);
+  }
+};
+
+class GcsTaskTable : public GcsTableWithIndex<TaskID, TaskTableData> {
+ public:
+  explicit GcsTaskTable(
+      std::shared_ptr<StoreClient<TaskID, TaskTableData, JobID>> store_client)
+      : GcsTableWithIndex(std::move(store_client)) {
+    table_name_ = TablePrefix_Name(TablePrefix::TASK);
+  }
+
+  Status BatchDelete(const std::vector<TaskID> &keys, const StatusCallback &callback);
```

Comment:
I think we can put it in the super class. Other tables would need batch delete as well.

Reply:
fixed